File: src/main/java/kookmin/software/capstone2023/timebank/domain/model/Account.java
package kookmin.software.capstone2023.timebank.domain.model;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.SQLDelete;
import org.hibernate.annotations.Where;

import java.time.LocalDateTime;
import java.util.Set;

@Entity
@Table(name = "account")
@Where(clause = "deleted_at IS NULL")
@SQLDelete(sql = "UPDATE account SET deleted_at = now() WHERE id = ?")
@Getter
@Setter
public class Account extends BaseTimeEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, updatable = false, length = 20)
    @Enumerated(EnumType.STRING)
    private AccountType type;

    @Column(nullable = false, updatable = true, length = 20)
    private String name;

    @Embedded
    private AccountProfile profile;

    @Column(nullable = true, updatable = true)
    private LocalDateTime deletedAt;

    @OneToMany(mappedBy = "account", fetch = FetchType.LAZY)
    private Set<User> users;

    @OneToMany(mappedBy = "account", fetch = FetchType.LAZY, cascade = CascadeType.ALL)
    private Set<BankAccount> bankAccounts;

    public void updateName(String name) {
        this.name = name;
    }

    public Account(AccountType type, AccountProfile profile, String name) {
        this.type = type;
        this.profile = profile;
        this.name = name;
    }
}

File: src/main/java/kookmin/software/capstone2023/timebank/domain/model/Account.kt
package kookmin.software.capstone2023.timebank.domain.model

import jakarta.persistence.CascadeType
import jakarta.persistence.Column
import jakarta.persistence.Embedded
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.OneToMany
import jakarta.persistence.Table
import org.hibernate.annotations.SQLDelete
import org.hibernate.annotations.Where
import java.time.LocalDateTime

@Entity
@Table(name = "account")
@Where(clause = "deleted_at IS NULL")
@SQLDelete(sql = "UPDATE account SET deleted_at = now() WHERE id = ?")
class Account(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long = 0,

    @Column(nullable = false, updatable = false, length = 20)
    @Enumerated(EnumType.STRING)
    val type: AccountType,

    @Column(nullable = false, updatable = true, length = 20)
    var name: String,

    @Embedded
    var profile: AccountProfile? = null,

    @Column(nullable = true, updatable = true)
    var deletedAt: LocalDateTime? = null,

    @OneToMany(
        mappedBy = "account",
        fetch = FetchType.LAZY,
    )
    val users: Set<User> = emptySet(),

    @OneToMany(
        mappedBy = "account",
        fetch = FetchType.LAZY,
        cascade = [CascadeType.ALL],
    )
    val bankAccounts: Set<BankAccount> = emptySet(),
) : BaseTimeEntity() {
    fun updateName(name: String) {
        this.name = name
    }
}
